### ISSUE TITLE
Support Windows ARM64

### ISSUE THREAD

**Issue by richlander** [dotnet/runtime#36699]
# Support Windows ARM64

We are in the process of adding support for [Windows ARM64](https://docs.microsoft.com/en-us/windows/arm/) as part of the .NET 5.0 project. As a result, you will be able to run Windows Forms, WPF and ASP.NET Core apps on Windows ARM64 devices, such as the [Surface Pro X](https://www.microsoft.com/en-us/p/surface-pro-x/8VDNRP2M6HHC).

We intend to support the same .NET capabilities on ARM64 as on x64. Related: https://github.com/dotnet/runtime/issues/35853

This project follows a similar effort to add Console and ASP.NET Core support for [Linux ARM64](https://github.com/dotnet/announcements/issues/82).

The [.NET 5.0 Preview 4 SDK](https://dotnet.microsoft.com/download/dotnet/5.0) supports building and running ASP.NET Core and Console apps.

You can install [`master` branch builds](https://github.com/dotnet/installer#installers-and-binaries) if you would like want to test Windows Forms on ARM64. `master` branch builds do not yet include WPF support for ARM64.

## Remaining .NET 5 Work

- [x] Windows ARM64 ASP.NET Core MSI (necessary for the SDK install): https://github.com/dotnet/aspnetcore/pull/25579
- [ ] Windows ARM64 SDK MSI: https://github.com/dotnet/installer/pull/8470

## Post .NET 5 Work

* [ ] Produce and publish Windows Desktop Runtime Pack for ARM64
  * [x] Add Windows Forms to the Windows Desktop Runtime Pack
  * [ ] Add WPF to the Windows Desktop Runtime Pack
* [ ] Add Windows Desktop framework to the SDK
  * [x] Add Windows Forms to the Windows Desktop framework in the SDK
  * [ ] Add WPF to the Windows Desktop framework in the SDK
  * [x] Crossgen Windows Desktop Pack for ARM64
* [ ] Test PowerShell on Windows ARM64 (PowerShell team will do this, but it is an important signoff criteria)

## Timeline

The following timeline is for .NET 5.0 and Windows ARM64:
*	.NET 5.0 Preview 4 -- Ship the .NET SDK, without Windows Desktop
*	.NET 5.0 Preview 5 -- Add Windows Desktop to SDK, with Windows Forms
*	.NET 5.0 Preview 6 -- Add ARM64 MSI installer for .NET Runtime 
* 	.NET 5.0 RC2 -- Add ARM64 MSI installer for SDK
* TBD -- Add WPF to Windows Desktop

After this project is complete, the Windows ARM64 SDK will have functional parity with x64.

## Examples

The following image demonstrate the [Conway's Game of life](https://github.com/dotnet/samples/tree/master/windowsforms/Conway's-Game-of-Life/VB) VB sample running on Windows ARM64.

<img width="398" alt="2020-05-15" src="https://user-images.githubusercontent.com/2608468/82086979-20f5bd00-96a4-11ea-8d73-abed8f2505fb.png">

The following demonstrates displaying [version information in a Windows Forms app](https://github.com/richlander/testapps/tree/master/versioninfo-windowsforms) on ARM64.

![image](https://user-images.githubusercontent.com/2608468/87205382-2cd4b880-c2bc-11ea-9c3f-a9d6f02b3af2.png)

The following image demonstrates 32-bit x86 Windows PowerShell 5.1 running emulated (on .NET Framework 4.8), and 64-bit ARM64 PowerShell 7.1 running natively (on .NET 5.0) on Windows ARM64.

<img width="375" alt="2020-05-15 (5)" src="https://user-images.githubusercontent.com/2608468/82128604-3dabf680-9771-11ea-9847-ca977549b21d.png">

See [PowerShell/PowerShell](https://github.com/powershell/powershell) to download the Windows ARM64 version. You can also install the [PowerShell .NET tool](https://www.nuget.org/packages/PowerShell/) on Windows ARM64.

See [versioninfo](https://github.com/richlander/testapps/blob/master/versioninfo/Program.cs) for APIs that you can use to get version and architecture information.

**Comment by JensNordenbro** [comment]
Hi!

I would like to request Windows IoT Enterprise to gain .NET5 support on Windows ARM64 (including Windows Forms, WPF, ..).

**Comment by voronoipotato** [comment]
Thanks Joel for referring to this issue :) I didn't know this existed.

**Comment by JensNordenbro** [comment]
@richlander ; It is not clear when to expect ASP.NET Core support on Windows ARM64. 
The description says "The .NET 5.0 Preview 4 SDK supports building and running ASP.NET Core and Console apps." however .NET 5 preview 6 as well as 7 renders the following message: 
C:\Program Files\dotnet\sdk\5.0.100-preview.6.20318.15\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(376,5): error NETSDK1082: There was no runtime pack for Microsoft.AspNetCore.App available for the specified RuntimeIdentifier 'win-arm64'.

https://github.com/dotnet/runtime/issues/36699 seems to be in line with what I am expecting but it would be nice to know which preview it is scheduled for! Thanks :)

**Comment by danmosemsft** [comment]
Closing as we have completed this for 5.0.

**Comment by danmosemsft** [comment]
@richlander anything left here - can we close it now?

**Comment by Alovchin91** [comment]
There are still no MSI installers available for Arm64. Is there a separate tracking issue for those? What is the plan to deliver the installers? Thanks!

**Comment by danmosemsft** [comment]
@dsplaisted @dagood can one of you answer that question?

**Comment by Alovchin91** [comment]
There is dotnet/installer#7985 but it refers back to this one. Circular dependency 😅

**Comment by dagood** [comment]
As written, this issue can't be closed yet. It's an epic that depends on https://github.com/dotnet/installer/issues/7985.
Thanks for pointing this out, @Alovchin91.

It's odd that the timeline contains ".NET 5.0 Preview 6 -- Add ARM64 MSI installers", but the "remaining work" section doesn't mention it. My assumption is this epic hasn't quite been kept up to date.

/cc @NikolaMilosavljevic

**Comment by danmosemsft** [comment]
Sounds good, glad I poked the issue then 😄

**Comment by danmosemsft** [comment]
Just one question @dagood who is the primary person responsible for driving dotnet/installer#7985 to resolution? Just so it doesn't get lost between teams, etc.

**Comment by dagood** [comment]
> Just one question @dagood who is the primary person responsible for driving [dotnet/installer#7985](https://github.com/dotnet/installer/issues/7985) to resolution?

@joeloff is currently assigned and there's active conversation there--I think that's accurate. 🙂

**Comment by Alovchin91** [comment]
@danmosemsft @dagood @richlander  Could someone please update and clarify the original post in this thread? As far as I understand, the plan has changed and there will be no Windows Desktop support for Arm64 in .NET 5?

**Comment by JensNordenbro** [comment]
Please do NOT remove stuff that worked in previous previews, i. e the ability to build for winforms for arm64 on an x64 development machine just because the sdk might not be available for arm64 development.

**Comment by Gerdya** [comment]
Looks like we have an ARM64 runtime installer now with .NET 5 Preview 8 

Looking forward to WPF support :)

**Comment by Alovchin91** [comment]
> Looks like we have an ARM64 runtime installer now with .NET 5 Preview 8 

@Gerdya  Where do you see it? It's not listed on the official website.

**Comment by JeffUnderhill** [comment]
@Alovchin91  See https://devblogs.microsoft.com/dotnet/announcing-net-5-0-preview-8/ and https://dotnet.microsoft.com/download/dotnet/5.0 - hope this helps.

**Comment by Alovchin91** [comment]
@JeffUnderhill  We are talking about Windows ARM64 installers, right? 😊

I might be missing something, but I can't find them on .NET 5.0 Preview 8 download page:

![image](https://user-images.githubusercontent.com/8490695/91314661-183e7800-e7b7-11ea-8621-96706a64f574.jpeg)

**Comment by dagood** [comment]
The release notes look inconsistent: https://github.com/dotnet/core/blob/master/release-notes/5.0/preview/5.0.0-preview.8.md#downloads

> ![image](https://user-images.githubusercontent.com/12819531/91315480-58810480-e77d-11ea-8bcc-8d88a33e9011.png)

IIRC the mechanisms for these tables are different and this is likely not intentional, but it's definitely confusing.

@rbhanda @leecow @mairaw ideas?

**Comment by rbhanda** [comment]
@NikolaMilosavljevic who helped add these entries for release notes

**Comment by NikolaMilosavljevic** [comment]
> @NikolaMilosavljevic who helped add these entries for release notes

I'm not sure how table is formatted - I've only provided the link to arm64 installer and asked for it to be added to release notes.

@dagood is the inconsistency with formatting or content?

**Comment by Alovchin91** [comment]
@NikolaMilosavljevic  It looks like the release notes page is missing ARM64 binaries link, while the official download page (https://dotnet.microsoft.com/download/dotnet/5.0) is missing ARM64 runtime installer link.

**Comment by NikolaMilosavljevic** [comment]
> @NikolaMilosavljevic It looks like the release notes page is missing ARM64 binaries link, while the official download page (https://dotnet.microsoft.com/download/dotnet/5.0) is missing ARM64 runtime installer link.

@rbhanda can these issues be fixed quickly?

**Comment by mairaw** [comment]
@rbhanda just pushed an update to the https://dotnet.microsoft.com/download/dotnet/5.0 but I'm not sure if the issue was resolved or not.

**Comment by Alovchin91** [comment]
@mairaw  No, the runtime installer link for Windows ARM64 is still absent.

**Comment by rbhanda** [comment]
@mairaw and I are looking into this. However the runtime installer is present in [release notes](https://github.com/dotnet/core/blob/master/release-notes/5.0/preview/5.0.0-preview.8.md)

**Comment by mairaw** [comment]
We pushed some fixes to the website yesterday and now we have the ARM64 entry for Windows runtime as well.

**Comment by ericstj** [comment]
@tommcdon can you update with any remaining work needed for this to be complete in 5.0?

**Comment by joeloff** [comment]
SDK is waiting on the ASP.NET changes - that PR is being reviewed. Once that's in we can start on the SDK installers.

**Comment by danmosemsft** [comment]
@joeloff it seems to me there is no longer value in having this meta-issue open. It's purpose was for the community to see what was going on during the cycle. You have individual issues for the work you mention, right? If you confirm that we can close this.

**Comment by tommcdon** [comment]
> @tommcdon can you update with any remaining work needed for this to be complete in 5.0?

- [x] Windows ARM64 ASP.NET Core MSI (necessary for the SDK install): https://github.com/dotnet/aspnetcore/pull/25579
- [ ] Port https://github.com/dotnet/aspnetcore/pull/25579 to RC2
- [ ] Windows ARM64 SDK MSI: https://github.com/dotnet/installer/pull/8470

**Comment by danmosemsft** [comment]
Shall we close this in favor of dotnet/installer#8470?

**Comment by tommcdon** [comment]
> Shall we close this in favor of [dotnet/installer#8470](https://github.com/dotnet/installer/pull/8470)?

Closing this issue as we have a pending PR that completes .NET 5.0 for Windows ARM64 support.

**Comment by Alovchin91** [comment]
Is there a tracking issue for WPF on Arm64 and adding back the Desktop Pack support? Would be nice to have links in the original post.

**Comment by fabiant3** [comment]
[](https://github.com/dotnet/wpf/pull/4048) is a recent pull request pointing at a few other ARM64 related issues. We are aiming to have WPF ARM64 support in the upcoming .NET6 Preview1.

**Comment by richlander** [comment]
Here is a tracking issue: https://github.com/dotnet/installer/issues/8905.
